sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/UnsafeHybridAggregationIterator.scala

```diff
@@ -0,0 +1,398 @@
+
+    val newValueAttributes =
+      allAggregateExpressions.flatMap(_.aggregateFunction.cloneBufferAttributes)
+
+    sortBasedAggregationIterator = SortBasedAggregationIterator.createFromKVIterator(
+      groupingKeyAttributes = groupingKeyAttributes,
+      valueAttributes = newValueAttributes,
+      inputKVIterator = sortedKVIterator.asInstanceOf[KVIterator[InternalRow, InternalRow]],
+      nonCompleteAggregateExpressions = newNonCompleteAggregateExpressions,
+      nonCompleteAggregateAttributes = newNonCompleteAggregateAttributes,
+      completeAggregateExpressions = Nil,
+      completeAggregateAttributes = Nil,
+      initialInputBufferOffset = 0,
+      resultExpressions = resultExpressions,
+      newMutableProjection = newMutableProjection,
+      outputsUnsafeRows = outputsUnsafeRows)
+  }
+
+  ///////////////////////////////////////////////////////////////////////////
+  // Methods used to initialize this iterator.
+  ///////////////////////////////////////////////////////////////////////////
+
+  /** Starts to read input rows and falls back to sort-based aggregation if necessary. */
+  protected def initialize(): Unit = {
+    var hasNext = inputKVIterator.next()
+    while (!sortBased && hasNext) {
+      val groupingKey = inputKVIterator.getKey()
+      val currentRow = inputKVIterator.getValue()
+      val buffer = buffers.getAggregationBuffer(groupingKey)
+      if (buffer == null) {
+        // buffer == null means that we could not allocate more memory.
+        // Now, we need to spill the map and switch to sort-based aggregation.
+        switchToSortBasedAggregation(groupingKey, currentRow)
+        sortBased = true
+      } else {
+        processRow(buffer, currentRow)
+        hasNext = inputKVIterator.next()
+      }
+    }
+  }
+
+  // This is the starting point of this iterator.
+  initialize()
+
+  // Creates the iterator for the Hash Aggregation Map after we have populated
+  // contents of that map.
+  private[this] val aggregationBufferMapIterator = buffers.iterator()
+
+  private[this] var _mapIteratorHasNext = false
+
+  // Pre-load the first key-value pair from the map to make hasNext idempotent.
+  if (!sortBased) {
+    _mapIteratorHasNext = aggregationBufferMapIterator.next()
+    // If the map is empty, we just free it.
+    if (!_mapIteratorHasNext) {
+      buffers.free()
+    }
+  }
+
+  ///////////////////////////////////////////////////////////////////////////
+  // Iterator's public methods
+  ///////////////////////////////////////////////////////////////////////////
+
+  override final def hasNext: Boolean = {
+    (sortBased && sortBasedAggregationIterator.hasNext) || (!sortBased && _mapIteratorHasNext)
+  }
+
+
+  override final def next(): InternalRow = {
+    if (hasNext) {
+      if (sortBased) {
+        sortBasedAggregationIterator.next()
+      } else {
+        // We did not fall back to the sort-based aggregation.
+        val result =
+          generateOutput(
```

Comment:
we should think about all of these function reuse -- i think they do make things more complicated, and they are bad for performance because there is nothing the compiler can do here about inlining them. 

it's actually more expensive than using if branches inline because the branch predictor usually does a pretty good job with fixed conditions for an operator (e.g. for a particular period of time, this operator will always be in a specific mode).